### ISSUE TITLE
Some fediverse clients don't rewrite hashtag links

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1634]
In the fediverse, hashtags in `content` are like mentions: we wrap them in HTML links, and when remote instances render the post, they rewrite those links to point to the hashtag search page on their instance. Gory details in #45.

...except evidently not all clients do this. The Mastodon Android app, for example, leaves them as is: https://piaille.fr/@politipet/113664287902122111

Bridgy Fed currently points its placeholder links in `content` to non-existent pages, eg #foo gets linked to https://bsky.brid.gy/hashtag/foo. I don't plan to build a hashtag or search UI in Bridgy Fed's web UI, and hashtags in content do need to be linked somewhere for clients that do rewrite, so I'm not sure what to do. Maybe link to `#`, ie the current page?

**Comment by snarfed** [comment]
> Maybe link to `#`, ie the current page?

Refreshed my memory a bit, and at least based on Mastodon's implementation back in March 2023, sounds like this wouldn't work. From https://github.com/snarfed/bridgy-fed/issues/45#issuecomment-1468856026 :

> * Both the tag object name and content link text must start with #.
> * Both the tag object and content hrefs must be non-empty, fully qualified URLs.
> * They don't need to match.
> * Mastodon rewrites the content href to point to the local instance's search for the hashtag.
> * However, it also generates a link preview for the original content href, unless it points to the sending instance. Ie for BF, if the content href is https://fed.brid.gy/[anything], the link preview is suppressed. That's not really useful for users though, they're not going to link their hashtags on their site to BF. 😐

**Comment by Tamschi** [comment]
I think these should link or be redirected to hashtag searches (not text searches for the hashtag, those are different there!) on Bluesky. It's essentially what all other fediverse instances do too.

**Comment by snarfed** [comment]
Good idea! Totally makes sense, and easy to do.

**Comment by snarfed** [comment]
^ Deploying a change to link Bluesky => fediverse converted hashtags to the bsky.app hashtag search page.

**Comment by Tamschi** [comment]
Hm, I think that might enable the link preview for them, though.
If that happens, I think it would be better to still have the https://bsky.brid.gy/hashtag/… URLs and HTTP-redirect to Bluesky from there.

**Comment by snarfed** [comment]
Hmm, receiving fediverse instances will generally rewrite hashtag links instead of generating previews for them, right? That's what they do now. Non-webapp clients like Mastodon Android may not do that rewriting, but afaik they also don't generate previews themselves locally...?

(In any case, I'm happy to do _some_ work like this to smooth over these links, but only some. I have kinda limited patience for interop depending on how finicky and niche it gets. 🤷)

**Comment by Tamschi** [comment]
> Hmm, receiving fediverse instances will generally rewrite hashtag links instead of generating previews for them, right? That's what they do now. Non-webapp clients like Mastodon Android may not do that rewriting, but afaik they also don't generate previews themselves locally...?
> 
> (In any case, I'm happy to do _some_ work like this to smooth over these links, but only some. I have kinda limited patience for interop depending on how finicky and niche it gets. 🤷)

Oh, if it works then I take it back of course 😅
It was just a hunch based on https://github.com/snarfed/bridgy-fed/issues/1634#issuecomment-2547358605.

**Comment by gerbenjacobs** [comment]
I'm a web user and I use absolute URLs to 'mysite/tags/the-tag-itself' as I have an actual page that's catching those.

Both those inline tags (from `e-content`) are properly shown in Mastodon and Bsky.app.

![Image](https://github.com/user-attachments/assets/da8f76b3-d410-4d13-afab-3edcfb56d5be)

Somehow Mastodon still takes the old original URL and uses that as link preview. Also any inline tags get removed from the added (`p-category`) tags. I had also added "#bluesky" but it's not shown in the bottom.

**Comment by snarfed** [comment]
@gerbenjacobs ruh roh! That's not good. Odd though, I'm not seeing that link preview on eg indieweb.social, on either desktop or mobile web. Where is your screenshot from?

![image](https://github.com/user-attachments/assets/7cdfa377-9bd8-4aa6-9601-e900eee77bd7)

**Comment by popet-master** [comment]
How about bsky.brid.gy/hashtag/foo simply returns a 302 and redirect to mastodon.social/hashtag/foo ?

**Comment by snarfed** [comment]
> How about https://bsky.brid.gy/hashtag/TAG simply returns a 302 and redirect to https://mastodon.social/tags/TAG ?

It's a nice idea, but it privileges mastodon.social a bit more than I'd like. bsky.app is very much the canonical Bluesky client app right now, but mastodon.social is much less the canonical fediverse instance, and fedi people often don't love seeing it enshrined as "special" like this compared to other instances.

**Comment by snarfed** [comment]
Oh, also, @gerbenjacobs, I misinterpreted your screenshot earlier, I think it's mostly doing the right/expected thing after all. `#bluesky` isn't shown at the bottom because it's visible in the post text, and its link in the post text is rewritten to the local fediverse instance's tag page.

I looked at it on three different instances - indieweb.social, mas.to, mastodon.social - and the last two do have the link preview you saw. Not sure why the difference there.

**Comment by popet-master** [comment]
use piaille.fr then :-) it's the biggest :fr: instance afaik, and "neutral". Maybe as a quick fix.

**Comment by gerbenjacobs** [comment]
> Oh, also, [@gerbenjacobs](https://github.com/gerbenjacobs), I misinterpreted your screenshot earlier, I think it's mostly doing the right/expected thing after all. `#bluesky` isn't shown at the bottom because it's visible in the post text, and its link in the post text is rewritten to the local fediverse instance's tag page.
> 
> I looked at it on three different instances - indieweb.social, mas.to, mastodon.social - and the last two do have the link preview you saw. Not sure why the difference there.

Yeah, I'm not sure there's much to do.

The screenshot was from todon.nl which is running Mastodon 4.3.2. I also have my own instance running the latest GotoSocial. 
The funny part (which I didn't even know) using a 'headless' client like Semaphore.social or Elk.zone will show both entries differently.

Those that are stored on todon.nl (running Mastodon) have this link preview, while those stored by GtS don't have those. GtS also doesn't do anything with hashtags and just keeps it a regular link back to my site.

And semaphore.social always uses the spoiler mechanic for that specific post with "Long post".

Bottomline: many servers and many clients do a lot of different things 🤷🏼‍♂️

**Comment by snarfed** [comment]
> I think it would be a good idea to `301 Moved Permanently`-redirect https://bsky.brid.gy/hashtag/TAG URLs to https://bsky.app/hashtag/TAG searches to fix older posts, since we don't send out `Update`s for post transform fixes like these.

Good idea, thanks, will do.

**Comment by snarfed** [comment]
Done! Should be deployed in a couple min.

**Comment by popet-master** [comment]
Issue with redirecting tag URLs to bsky.app (instead of redirecting to some mastodon instance) is that now, the user experience with clicking hashtag links is "confusing" for Mastodon users: if I click on a hashtag link in a bridged post, from the Mastodon Web UI, it gets me to the Mastodon tag page (as expected), but if I click the same link from the Mastodon Android app, it gets me to the Bluesky web page (i.e. in an "external" browser view, not in-app view) which is 1./ slow and 2./ "not so consistent" (as I would expect to "stay" in Mastodon, as per the Web UI)

**Comment by Tamschi** [comment]
Parsing difference, I believe. Mastodon's Android app [really wants a `hashtag` `class` on the link](https://github.com/mastodon/mastodon-android/blob/9361c6d0dfcd1eb9d641f86219df6f5a86e44dac/mastodon/src/main/java/org/joinmastodon/android/ui/text/HtmlParser.java#L148), and Bridgy Fed currently doesn't provide that. The web version doesn't require it.

It will likely start working for new posts if we add the `class` to the `<a>` tag, but I'll file an issue regarding the parsing inconsistency with Mastodon's Android app.

**Comment by snarfed** [comment]
@popet-master yes! You're absolutely right, I'd much rather hashtags always link internally, like they do in most web-based fediverse UIs, but we haven't yet found a reliable way to do that in all apps.

@Tamschi good find! Easy to add that class, I'll try now.

**Comment by snarfed** [comment]
Done, should be deployed in ~15m. @popet-master feel free to try with a new post after that!

**Comment by snarfed** [comment]
@popet-master friendly nudge, any luck here?

**Comment by popet-master** [comment]
Thanks for the nudge @snarfed ! Sorry I missed your first ping 3 weeks ago. Eventually I installed Moshidon on my Android mobile, as it seemed to "work" better than the official Mastodon app re: hashtag links. But it might have been due to your update :-)

I'll re-test on bot old and new posts with both Moshidon and the Mastodon app - and edit this answer with findings

**Comment by snarfed** [comment]
Thanks for testing again @popet-master! And ugh, I see the remaining bug on our side. Fixing now.

**Comment by snarfed** [comment]
Looks like that last bug fix ^ did the trick! I posted a new post, https://fed.brid.gy/r/https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3lhkmgz6xnc2h , opened it in the Mastodon app and tapped on a hashtag, and it opened the hashtag inside the app. 🎉

**Comment by Daft-Freak** [comment]
This does seem to have caused mastodon to generate link previews for tags...

(A recent example I've seen: https://fosstodon.org/@arturo182.bsky.social@bsky.brid.gy/114002439374525040)

**Comment by snarfed** [comment]
Ugh yeah true. Doesn't seem due to this latest change though, I think that's been happening for a while, eg https://fosstodon.org/@arturo182.bsky.social@bsky.brid.gy/113928335037043303

I've been jumping through hoops trying to match Mastodon's undocumented req'ts for clean @-mentions and hashtags for years and years now, I'm running low on patience for reverse engineering many more hoops.

**Comment by Daft-Freak** [comment]
Possibly happening since the change to link tags to bluesky, though it does seem silly that Mastodon is rewriting the link but also using the old one to generate a preview...

**Comment by snarfed** [comment]
Agreed. I wonder if it suppresses the preview for links to the originating instance's domain. Guess I could try that.
